### ISSUE TITLE
Visual Predictive Check

### ISSUE THREAD

**Issue by wds15** [stan-dev/bayesplot#66]
It would be great to have a so-called visual predictive check. To exemplify it I include an example in the form of a simple example.

This plot is very useful for models which have continuous regressors which are given by the design of the experiemnt at the same value for all the subjects in a data-set. For example, imagine I have many subjects in a clinical study and one measures at pre-defined time-points for all patients whatever is of interest. The plot then allows to compare the raw quantiles of the data at each time-point vs what the model predicts for these quantiles (with its uncertainty).

[vpc_example_R.txt](https://github.com/stan-dev/bayesplot/files/732945/vpc_example_R.txt)

[vpc_mtcars.pdf](https://github.com/stan-dev/bayesplot/files/732949/vpc_mtcars.pdf)

**Comment by dpastoor** [comment]
@wds15 are you aware of https://github.com/ronkeizer/vpc

**Comment by jgabry** [comment]
Don't know about @wds15, but I hadn't seen that package. Thanks.

Although the vpc suggested here seems easy enough to add, so we might as
well do our own version to maintain a consistent style across plot types
and also limit dependencies. Especially since one of the goals for
bayesplot is to allow other packages to call it in order to not have to
 use different packages and styles for each kind of plot.

**Comment by dpastoor** [comment]
yeah, the basic functionality is really easy to implement, and could be nice stylistically.

The 3 'extras' this package offers are:

* binning for non-uniform/nominal sampling times
* prediction corrections - useful when data has wide range, eg ascending dose studies (more details here: https://www.ncbi.nlm.nih.gov/pubmed/21302010)
* VPC's for categorical data

Happy to answer any questions, or help in some capacity. I think a 'built-in' VPC would be a valuable addition, and is a go-to plot in the pharmacometrics community.

**Comment by jgabry** [comment]
Cool, thanks. I don't do much pharma stuff so it will definitely be good to
get some feedback on the plots when I have a chance to add it to a branch.

**Comment by wds15** [comment]
@dpastoor no, I don't know that package, looks interesting.

However, having this plot in bayesplot would be great for the reasons Jonah raises. The plot is one of the standards in PK/PD modeling. However, I think this is useful elsewhere certainly. Given with what is there in bayesplot it should be straightforward; at least it looks like.

I am happy to review the feature branch (or even write one should I think adapting code from other plotting functions looks easily adaptable to me in a coherent design).

**Comment by jgabry** [comment]
@wds15 In the example you sent which parts of the data processing (e.g. binning) do you envision the user performing before calling a bayesplot function?

**Comment by jgabry** [comment]
Also, we should find a good way to write the code without using the plyr package. dplyr and reshape2 are ok since bayesplot already uses them.

**Comment by wds15** [comment]
The user should be doing the binning himself. bayesplot should only need to know what refers to the bins. I suppose this can be mapped to the grouping facilities which are already there.

What would be useful is to allow the user to specify extra strata which are used to split things, e.g. by facets. So the bins always define the x-axis and a variable like sex may put male and female subjects on different facets.

Let me know if I should look at some branch and try things.

**Comment by jgabry** [comment]
Ok cool. I think that makes sense. I'll try to put together a prototype soon.

**Comment by utkarshpawade** [comment]
hey @jgabry 
VPC-style plots are still the missing piece for anyone using bayesplot on longitudinal / pharmacometric / time-indexed data, and I think the groundwork that's landed since 2017 (specifically `ppc_intervals()` / `ppc_ribbon()` with an `x` argument, and `ppc_intervals_data()`) makes a clean implementation much cheaper now than it would have been originally.

Technical approach: A VPC is, structurally, three overlaid `ppc_ribbon()`-shaped objects: the empirical quantile trajectories of `y` at each `x`-bin, and the posterior median + credible band of each of those quantiles computed across `yrep`.

Concretely:

- Add `ppc_vpc()` / `ppc_vpc_data()` in a new `R/ppc-vpc.R`, following the file/function conventions of `R/ppc-intervals.R`.
- Reuse `ppc_intervals_data(y, yrep, x, group)` as the tidy back end, but aggregate within `x`-bins (user-supplied breaks or `pretty()`-based default) and compute a configurable set of quantiles (default `c(0.1, 0.5, 0.9)` per the Karlsson/Savic convention). For each quantile, compute its posterior median and `prob`/`prob_outer` interval across draws, the same data shape `ppc_ribbon()` already consumes.
- Rendering is then `geom_ribbon()` + `geom_line()` per quantile, with `bayesplot_theme_get()` + `scale_color_ppc()`. The user-facing argument list mirrors `ppc_intervals()` almost exactly, which keeps the API learnable.
- Stratified VPC (per-covariate-bin) falls out of the `group` argument that `ppc_intervals_grouped()` already uses, no new faceting machinery needed.

Edge cases:
Binning over a continuous `x`: For discretely-dosed designs `x` is already categorical and grouping is trivial; for continuous `x` the user needs control over bins. I'd expose a `bins` argument (numeric vector of breaks, an integer count, or `NULL` for `pretty()`), and document that the observed-quantile curves are only interpretable when each bin has a reasonable `n`.

Prediction-corrected VPC (pcVPC): The literature treats pcVPC as the more defensible variant for models with covariate-driven between-subject variability. I'd leave it out of the first PR (it requires the predicted typical value per observation, which isn't part of the `(y, yrep)` contract) but structure `ppc_vpc_data()` so a future `pred` argument can slot in without an API break.

---

Before I prototype anything, would you want this surfaced as `ppc_vpc()` in the main PPC family, or as a thinner wrapper around `ppc_ribbon()` with a `quantiles =` argument? The latter is less code but harder to document; the former is discoverable via `help(package = "bayesplot")`.
